File: ios/StockbookTests/ThemeTests.swift
import Testing
import Foundation
@testable import Stockbook

/// The theme is a preference of the phone, not a fact about the shop.
///
/// That makes it behave like the language and unlike the currency, and the rules
/// below are the ones that difference produces. None of them can be checked by
/// looking at a screen — a theme that quietly resets itself looks like a screen
/// somebody drew wrong.
@MainActor
@Suite("Theme")
struct ThemeTests {

    private func makeStore() -> StockbookStore {
        StockbookStore(repository: InMemoryRepository())
    }

    /// `Nocturne` is global, so a test that leaves it on light would hand the
    /// next one a palette it never asked for.
    private func restoringTheme(_ work: () throws -> Void) rethrows {
        defer { Nocturne.use(.dark) }
        try work()
    }

    @Test("A shop opens dark until somebody says otherwise")
    func defaultsToDark() {
        #expect(makeStore().settings.theme == .dark)
        #expect(Nocturne.bg == Palette.dark.bg)
    }

    @Test("Choosing a theme persists it and takes effect immediately")
    func setTheme() throws {
        try restoringTheme {
            let repository = InMemoryRepository()
            let store = StockbookStore(repository: repository)

            store.setTheme(.light)

            let onDisk = try repository.loadAll().settings.theme

            #expect(store.settings.theme == .light)
            #expect(onDisk == .light, "not written to disk")
            #expect(Nocturne.bg == Palette.light.bg, "the palette in force did not follow the setting")
        }
    }

    /// A default value does **not** make Swift's synthesised decoder tolerate a
    /// missing key — the reason `Settings` decodes by hand, and the reason this
    /// test exists for every field added after v1.
    @Test("A shop file written before themes existed still opens, in the dark")
    func decodesWithoutTheKey() throws {
        let json = Data("""
        { "ownerName": "Khalid Al-Amri", "currencyCode": "SAR", "setupCompleted": true }
        """.utf8)

        let settings = try JSONDecoder().decode(Settings.self, from: json)

        #expect(settings.theme == .dark)
        #expect(settings.ownerName == "Khalid Al-Amri")
    }

    @Test("Starting over keeps the theme")
    func startOverKeepsTheme() {
        restoringTheme {
            let store = makeStore()
            store.setTheme(.light)
            store.setOwnerName("Khalid")

            store.startOver()

            #expect(store.settings.ownerName.isEmpty)
            #expect(store.settings.theme == .light,
                    "wiping the shop is a data decision, not a decision about how it looks")
        }
    }

    @Test("Importing another phone's shop does not repaint this one")
    func importKeepsThisPhonesTheme() {
        restoringTheme {
            let store = makeStore()
            store.setTheme(.light)

            store.replaceEverything(with: BackupDocument(
                exportedAt: .now,
                ownerName: "Someone Else",
                currencyCode: "SAR",
                products: [],
                bills: []
            ))

            #expect(store.settings.theme == .light)
            #expect(store.settings.ownerName == "Someone Else", "the shop itself did come from the file")
        }
    }
}

File: android/core/src/test/kotlin/com/stockbook/core/ThemeTests.kt
package com.stockbook.core

import com.stockbook.core.model.AppTheme
import com.stockbook.core.model.ShopState
import com.stockbook.core.store.InMemoryRepository
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.AppLanguage
import com.stockbook.core.transfer.BackupDocument
import com.stockbook.core.transfer.BackupService
import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The theme is a preference of the phone, not a fact about the shop.
 *
 * Which makes it behave like the language and unlike the currency, and the three
 * rules below are the ones that difference produces. None of them can be seen by
 * looking at a screen — a theme that resets itself looks like a screen somebody
 * drew wrong.
 */
class ThemeTests {

    private fun store() = StockbookStore(InMemoryRepository())

    @Test
    fun `a shop opens dark until somebody says otherwise`() {
        assertEquals(AppTheme.DARK, store().settings.theme)
    }

    @Test
    fun `choosing a theme persists it`() {
        val repository = InMemoryRepository()
        val store = StockbookStore(repository)

        store.setTheme(AppTheme.LIGHT)

        assertEquals(AppTheme.LIGHT, store.settings.theme)
        assertEquals(AppTheme.LIGHT, repository.loadAll().settings.theme, message = "not written to disk")
    }

    @Test
    fun `a shop file written before themes existed still opens, in the dark`() {
        val before = """
            {
              "settings": { "ownerName": "Khalid Al-Amri", "currencyCode": "SAR", "setupCompleted": true }
            }
        """.trimIndent()

        val state = BackupService.json.decodeFromString<ShopState>(before)

        assertEquals(AppTheme.DARK, state.settings.theme)
        assertEquals("Khalid Al-Amri", state.settings.ownerName)
    }

    @Test
    fun `starting over keeps the theme`() {
        val store = store()
        store.setTheme(AppTheme.LIGHT)
        store.setOwnerName("Khalid")

        store.startOver()

        assertTrue(store.settings.ownerName.isEmpty())
        assertEquals(
            AppTheme.LIGHT,
            store.settings.theme,
            message = "wiping the shop is a data decision, not a decision about how it looks"
        )
    }

    @Test
    fun `importing another phone's shop does not repaint this one`() {
        val store = store()
        store.setTheme(AppTheme.LIGHT)
        store.setLanguage(AppLanguage.KANNADA)

        store.replaceEverything(
            BackupDocument(
                exportedAt = Instant.parse("2026-08-01T09:00:00Z"),
                ownerName = "Someone Else",
                currencySymbol = "SAR ",
                products = emptyList(),
                bills = emptyList()
            )
        )

        assertEquals(AppTheme.LIGHT, store.settings.theme)
        assertEquals(AppLanguage.KANNADA, store.settings.language)
        assertEquals("Someone Else", store.settings.ownerName, message = "the shop itself did come from the file")
    }
}
